packages/ui/components/VisualEditor/SchemaObject.tsx

```diff
@@ -22,27 +22,88 @@ const SchemaObject: React.FC<SchemaObjectProps> = ({
+    console.log("propertyPath: ",propertyPath)
+    _.unset(updatedSchema, normalizedPath);
+    onSchemaChange(updatedSchema);
   };
 
-  const handleTypeChange = (propertyPath: string, newSchema: any) => { // Added types to resolve TS7006
-    console.log(`handleTypeChange called with path: ${propertyPath}, newType: ${newSchema}`);
+  const handleTypeChange = (propertyPath: string, newSchema: any, newType: any) => { // Added types to resolve TS7006
+    console.log(`handleTypeChange called with path: ${propertyPath}, newType: ${newType}`);
+    const normalizedPath = propertyPath.startsWith('.') ? propertyPath.slice(1) : propertyPath;
     const currentSchema = _.cloneDeep(schema);
-    _.set(currentSchema, propertyPath, newSchema);
+    const typePath = `${normalizedPath}.type`;
+
+    if(newType.type == "array") {
+      const itemType = newType.items;
+      console.log("itemType",itemType)
+      _.set(currentSchema, typePath, 'array');
+      _.set(currentSchema, `${normalizedPath}.items`, itemType);
+    } else {
+      const newTypeValue = newType.type;
+      _.set(currentSchema, typePath , newTypeValue);
+      const itemsPath = `${normalizedPath}.items`;
+      _.unset(currentSchema, itemsPath);
+    }
+
     console.log(`Type changed at ${propertyPath}`, newSchema);
     onSchemaChange(currentSchema);
   };
 
+  const handleToggleRequired = (path:string, name: string) => {
+    const updatedSchema = _.cloneDeep(schema);
+    const existingRequired = _.get(updatedSchema, `required`, []);
+    const isRequirePresent = existingRequired.includes(name);
+    console.log("isRequired",isRequirePresent);
+    if(!isRequirePresent) {  
+      const newRequired = _.uniq([...existingRequired, name]);
+      _.set(updatedSchema, `required`, newRequired);
+    } else {
+      const newRequired = existingRequired.filter((item: string) => item !== name);
+      _.set(updatedSchema, `required`, newRequired);
+    }
+  
+    onSchemaChange(updatedSchema);
+  }
+  
+  const handleToggleNestedRequired = (path: string, name: string) => {
+    const updatedSchema = _.cloneDeep(schema);
+    const normalizedPath = path.startsWith('.') ? path.slice(1) : path;
+    const schemaPath = normalizedPath.split('.properties');
+    console.log("schemaPath",schemaPath)
+    const requiredPath = schemaPath.slice(0, -1).join('.properties') + '.required';
+    console.log("requiredPath",requiredPath)
+  
+    const existingRequired = _.get(updatedSchema, requiredPath, []);
+    const isRequirePresent = existingRequired.includes(name);
+  
+    if (!isRequirePresent) {
+      const newRequired = _.uniq([...existingRequired, name]);
+      _.set(updatedSchema, requiredPath, newRequired);
+    } else {
+      const newRequired = existingRequired.filter((item: string) => item !== name);
+      _.set(updatedSchema, requiredPath, newRequired);
+    }
+  
+    onSchemaChange(updatedSchema);
+  };
+
   return (
     <div style={{ margin: '10px 0' }}>
```

Comment:
we can use https://www.radix-ui.com/primitives/docs/components/collapsible to collapse the Objects and arrays.

<img width="854" alt="Screenshot 2024-05-20 at 12 20 48" src="https://github.com/asyncapi/studio/assets/32505158/48650273-4152-4539-ac25-c21b99235b40">